### PR TITLE
[RISCV] Add sifive-p670 processor

### PR DESCRIPTION
**Pull request by michaelmaitland**
This is an OOO core that has a vector unit. For more information see https://www.sifive.com/cores/performance-p650-670.

Scheduler model and other tuning will come in separate patches.

### CODE REVIEW

File: llvm/lib/Target/RISCV/RISCVProcessors.td [hunk 1]
```diff
@@ -237,6 +237,43 @@ def SIFIVE_P450 : RISCVProcessorModel<"sifive-p450", SiFiveP400Model,
...
+                                       FeatureStdExtZvl128b,
+                                       FeatureStdExtZvbb,
+                                       FeatureStdExtZvknc,
+                                       FeatureStdExtZvkng,
+                                       FeatureStdExtZvksc,
+                                       FeatureStdExtZvksg,
+                                       FeatureFastUnalignedAccess],
+                                      [TuneNoDefaultUnroll,
+                                       TuneConditionalCompressedMoveFusion,
+                                       TuneLUIADDIFusion,
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVProcessors.td:274]: This should have AUIPC fusion like p450

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 2]
```diff
@@ -2000,6 +2000,14 @@ bool RISCVTargetLowering::shouldSinkOperands(
   if (!I->getType()->isVectorTy() || !Subtarget.hasVInstructions())
     return false;
 
+  // Don't sink splat operands if the target prefers it. Some targets requires
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:2003]: This change is not mentioned in the patch description

File: llvm/lib/Target/RISCV/RISCVProcessors.td (hunk 1 above)
Comment (michaelmaitland) [reply llvm/lib/Target/RISCV/RISCVProcessors.td:274]: Updated.

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp (hunk 2 above)
Comment (michaelmaitland) [reply llvm/lib/Target/RISCV/RISCVISelLowering.cpp:2003]: Updated.

**Comment by topperc** [comment]
Need to update docs/ReleaseNotes.rst

File: clang/test/Driver/riscv-cpus.c [hunk 3]
```diff
@@ -241,7 +241,17 @@
 // MCPU-SIFIVE-P450-SAME: "-target-feature" "+zbb"
 // MCPU-SIFIVE-P450-SAME: "-target-feature" "+zbs"
 // MCPU-SIFIVE-P450-SAME: "-target-abi" "lp64d"
-//
+
+// RUN: %clang -target riscv64 -### -c %s 2>&1 -mcpu=sifive-p670 | FileCheck -check-prefix=MCPU-SIFIVE-P670 %s
```
Comment (wangpc-pp) [inline clang/test/Driver/riscv-cpus.c:245]: One feature one line?

File: llvm/lib/Target/RISCV/RISCVFeatures.td [hunk 4]
```diff
@@ -1082,6 +1082,13 @@ def TuneShortForwardBranchOpt
 def HasShortForwardBranchOpt : Predicate<"Subtarget->hasShortForwardBranchOpt()">;
 def NoShortForwardBranchOpt : Predicate<"!Subtarget->hasShortForwardBranchOpt()">;
 
+// P670 requires a S2V transfer buffer to move scalars into vectors.
+// FIXME: Forming .vx/.vf can reduce register pressure.
+def TuneDontSinkSplatOperands
```
Comment (wangpc-pp) [inline llvm/lib/Target/RISCV/RISCVFeatures.td:1087]: This can be split into another PR.

File: clang/test/Driver/riscv-cpus.c (hunk 3 above)
Comment (michaelmaitland) [reply clang/test/Driver/riscv-cpus.c:245]: updated.

File: llvm/lib/Target/RISCV/RISCVFeatures.td (hunk 4 above)
Comment (michaelmaitland) [reply llvm/lib/Target/RISCV/RISCVFeatures.td:1087]: https://github.com/llvm/llvm-project/pull/79199

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp (hunk 2 above)
Comment (michaelmaitland) [reply llvm/lib/Target/RISCV/RISCVISelLowering.cpp]: I took the changes out of this patch. I will work on the next patch once this lands since there are a few fixups here and it will be easier to work off the squashed commit.

**Review by topperc — APPROVED** [review]
LGTM

**Review by wangpc-pp — APPROVED** [review]
LGTM.
